source/main.cpp

```diff
@@ -131,32 +131,24 @@ bool init(int argc, char *argv[])
  */
 void shutdown()
 {
-    featuresReadWriteLock.lock(); // LOCK
-    // Make a copy of the features vector for thread safety
-    vector<Feature *> featuresCopy = features;
-    featuresReadWriteLock.unlock(); // UNLOCK
-
-    if (!attemptingShutdown && !featuresCopy.empty())
+    LOG_DEBUG(TAG, "Inside of shutdown()");
+    if (!attemptingShutdown && features->getSize() != 0)
     {
         attemptingShutdown = true;
 
-        for (auto &feature : featuresCopy)
-        {
-            LOGM_DEBUG(TAG, "Attempting shutdown of %s", feature->getName().c_str());
-            feature->stop();
-        }
-
-        resourceManager->dumpMemTrace();
+        LOG_DEBUG(TAG, "Calling stop all");
+        features->stopAll();
```

Comment:
Why not just remove each feature after you stop it? Seems like these are only called together.

Reply:
There were some logic issues doing it that way, which is how it was originally done. the call to remove requires the lock as well, recursive locks could be used in that case, but I was advised against them. I had issues with the existing logic inside of included handle_feature_stopped, which would call shutdown once the vector was empty which means we couldn't remove the last feature since now we track the features in a vector in FeatureRegistry. However, now that I removed that logic, we could just move the logic of this removeAll() call inside of stopAll(), but I don't see a need to currently. 
To summarize, I made it this way to work around other errors in main.cpp, but those errors are gone and now this remains, but I think keeping it this way is fine.

Reply:
Adding an entire new method to a class because of something that no longer exists is generally not recommended.